Framework/API/src/WorkspaceGroup.cpp

```diff
@@ -278,21 +279,35 @@ void WorkspaceGroup::workspaceDeleteHandle(
 }
 
 /**
- * Callback when a after-replace notification is received
- * Replaces a member if it was replaced in the ADS.
- * @param notice :: A pointer to a workspace after-replace notificiation object
+ * Callback when a before-replace notification is received
+ * Replaces a member if it was replaced in the ADS and checks
+ * for duplicate members within the group
+ * @param notice :: A pointer to a workspace before-replace notification object
  */
-void WorkspaceGroup::workspaceReplaceHandle(
+void WorkspaceGroup::workspaceBeforeReplaceHandle(
     Mantid::API::WorkspaceBeforeReplaceNotification_ptr notice) {
   std::lock_guard<std::recursive_mutex> _lock(m_mutex);
 
-  const std::string replacedName = notice->objectName();
+  const auto oldObject = notice->oldObject();
+
   for (auto &workspace : m_workspaces) {
```

Comment:
If we had an additional `else if(workspace == newObject)` and then stored an iterator to that position we could erase that single position after the loop? We'd need to take the `break` out but it would at least avoid more allocations by having to loop over and create a set each time.